### PR TITLE
Fix #267: openshift-maven-plugin does not update Routes

### LINKED ISSUE #267: openshift-maven-plugin does not update Routes
### Description
I originally found this issue in FMP fabric8:apply goal, I have yet to use the new plugin but just by looking at the code I can see the issue still exists in JKube.

The plugin will only create a Route if it doesn't exist and do nothing otherwise, here is the corresponding code in FMP:
https://github.com/fabric8io/fabric8-maven-plugin/blob/master/core/src/main/java/io/fabric8/maven/core/service/ApplyService.java#L668-L690
And the same code in JKube:
https://github.com/eclipse/jkube/blob/master/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/ApplyService.java#L646-L668

### Info

* Maven version (`mvn -v`) :
```
Apache Maven 3.6.2 (40f52333136460af0dc0d7232c0dc0bcf0d9e117; 2019-08-27T15:06:16Z)
Java version: 1.8.0_242, vendor: AdoptOpenJDK
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "3.10.0-1127.13.1.el7.x86_64", arch: "amd64", family: "unix"
```
* Kubernetes / Red Hat OpenShift setup and version :
OpenShift 4.3

* If it's a bug, how to reproduce :
1. Create a project with a Route resource fragment.
2. Deploy to OpenShift ```mvn oc:deploy```.
3. Make any change to the Route resource fragment.
4. Redeploy.
5. ```oc describe route <route>```, not updated.

### PR DESCRIPTION
**Pull request by mtrejo**
Signed-off-by: Marcos Trejo <marcos.trejo@gmail.com>

## Description
Generating a patch for Route, in a similar fashion as the plugin already did for Service and other resources. This fixes https://github.com/eclipse/jkube/issues/267

## Type of change
 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [x] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [x] I have implemented unit tests to cover my changes
 - [ ] ~I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly~
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [ ] ~I tested my code in Kubernetes~
 - [x] I tested my code in OpenShift

### CODE REVIEW

File: jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/ApplyService.java [hunk 1]
```diff
@@ -939,6 +926,17 @@ public void applyService(Service service, String sourceName) throws Exception {
         }
     }
 
+    private <T extends HasMetadata> void doPatchEntity(T oldEntity, T newEntity, String namespace, String sourceName) {
+        String kind = newEntity.getKind();
+        log.info("Updating {} from {}", kind, sourceName);
```
Comment (rohanKanojia) [inline jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/ApplyService.java:931]: I'm not sure `KitLogger` supports format specifiers

Comment (manusa) [reply jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/ApplyService.java:931]: It does, but using String Formatter placeholders, use following `log.info("Updating %s from %s", kind, sourceName);

File: jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Test
+    public void testCreateRoute() throws Exception {
+        Route route = buildRoute();
+
+        WebServerEventCollector<OpenShiftMockServer> collector = new WebServerEventCollector<>(mockServer);
+        mockServer.expect().get()
+                .withPath("/apis/route.openshift.io/v1/namespaces/default/routes/route")
+                .andReply(collector.record("get-route").andReturn(404, ""))
```
[the remark this answers is not in the document]
Comment (mtrejo) [reply jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java]: Updated

File: jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                .always();
+        mockServer.expect().post()
+                .withPath("/apis/route.openshift.io/v1/namespaces/default/routes")
+                .andReply(collector.record("new-route").andReturn(201, route))
```
[the remark this answers is not in the document]
Comment (mtrejo) [reply jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java]: Updated

File: jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Test
+    public void testCreateRouteInServiceOnlyMode() throws Exception {
+        Route route = buildRoute();
+
+        WebServerEventCollector<OpenShiftMockServer> collector = new WebServerEventCollector<>(mockServer);
+        mockServer.expect().get()
+                .withPath("/apis/route.openshift.io/v1/namespaces/default/routes/route")
+                .andReply(collector.record("get-route").andReturn(404, ""))
```
[the remark this answers is not in the document]
Comment (mtrejo) [reply jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java]: Updated

File: jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Test
+    public void testCreateRouteNotAllowed() throws Exception {
+        Route route = buildRoute();
+
+        WebServerEventCollector<OpenShiftMockServer> collector = new WebServerEventCollector<>(mockServer);
+        mockServer.expect().get()
+                .withPath("/apis/route.openshift.io/v1/namespaces/default/routes/route")
+                .andReply(collector.record("get-route").andReturn(404, ""))
```
[the remark this answers is not in the document]
Comment (mtrejo) [reply jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java]: Updated

File: jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
+package org.eclipse.jkube.kit.config.service;
```
[the remark this answers is not in the document]
Comment (mtrejo) [reply jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java:15]: Updated

File: jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/ApplyService.java (hunk 1 above)
Comment (mtrejo) [reply jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/ApplyService.java]: Thanks for your suggestions, I implemented them all.

**Comment by manusa** [comment]
Thanks for contributing :heart:

**Pull request merged**